Add Pipelines/Tekton Configuration Script (#7)
redhat-ai-dev/ai-rhdh-installer: docs/GITOPS-CONFIG.md

### AI-RHDH-Installer: Manual Configuration

#### Step 1: ConfigMaps and Secrets
In your chosen namespace you should apply the `argocd-config` ConfigMap under [`argocd-config.yaml` ](../resources/argocd-config.yaml).

> As part of the `ai-rhdh-installer` a secret was created in your desired namespace with the name `rhdh-argocd-secret`, keep note of this Secret as well as the two ConfigMaps applied above.

Reviewer: ```suggestion
As part of the `ai-rhdh-installer` a secret was created in your desired namespace with the name `rhdh-argocd-secret`, keep note of this Secret as well as the ConfigMap applied above.
```

Reviewer: Here I think it gets confusing. Installing with `ai-rhdh-installer` is mentioned as pre-requisite for the script configuration. Reading this sentence makes me think it should be for both right?

Author: > Here I think it gets confusing. Installing with `ai-rhdh-installer` is mentioned as pre-requisite for the script configuration. Reading this sentence makes me think it should be for both right?

@thepetk I'm not sure I fully understand the question, `ai-rhdh-installer` would be a prerequisite for all configure scripts as the configure scripts are ran after the fact, the only exception this is not the case is if the user already brought their own components in which case they need to follow the 'Pre-Existing Instance' steps. Does this touch on what you're asking?

Reviewer 2: Yeah I believe there might have been some confusion, in this case it is the manual configuration for tekton if you installed the resources using the `ai-rhdh-installer` so it would be a prereq, like @michael-valdron said the `pre-existing instances` would have other options available!

Reviewer: @michael-valdron @Jdubrick sharing some more context as my comment might be missleading. Basically, what I mean is:

* [### AI-RHDH-Installer: Script Configuration](https://github.com/redhat-ai-dev/ai-rhdh-installer/blob/b2898c501fd7b26a578ec9cdd15f7da7c58655bf/docs/GITOPS-CONFIG.md?plain=1#L17-L18) we explicitly mention that the user has to use `ai-rhdh-installer`.

* Tho in [## AI-RHDH-Installer: Manual Configuration](https://github.com/redhat-ai-dev/ai-rhdh-installer/blob/b2898c501fd7b26a578ec9cdd15f7da7c58655bf/docs/GITOPS-CONFIG.md?plain=1#L20-L25) things are more implicit (at least this is my impression from my first pass of the guide).

I think it might be more clear if we mention in the beggining of both ways (manual, script) that the `ai-rhdh-installer` is a requirement, like we do in the [list of contents](https://github.com/redhat-ai-dev/ai-rhdh-installer/blob/b2898c501fd7b26a578ec9cdd15f7da7c58655bf/docs/GITOPS-CONFIG.md?plain=1#L4-L9). WDYT?

Reviewer 2: In the wording of the content I suppose you're right but the sections are prefixed with the method, **`AI-RHDH-Installer`**, no?

#### Step 2: Updating Plugins
**Note: RHDH will encounter errors if the ArgoCD plugins are loaded before an instance is properly attached - This step may result in errors until all steps are completed**

To include the [ArgoCD plugins list](../dynamic-plugins/argocd-plugins.yaml) we need to edit the dynamic plugins ConfigMap that was created by the RHDH Operator:

![Dynamic Plugins Example](../assets/dynamic-plugins-example.png)

Edit the associated `yaml` file to include the contents of the [ArgoCD plugins list](../dynamic-plugins/argocd-plugins.yaml) under the `plugins` section:

![Dynamic Plugins Example 2](../assets/dynamic-plugins-example-2.png)

> Alternatively, we can use this series of commands to perform the same task with `kubectl` and `yq` using the [`argocd-plugins.yaml`](../dynamic-plugins/argocd-plugins.yaml):

Reviewer: Might be more clear here if we have 2 sub-sections:

* Edit the associated `yaml` file to include..
* OR Alternatively, we can use this series...

Author: Also done in https://github.com/redhat-ai-dev/ai-rhdh-installer/pull/7/commits/143b26952fac4d4be9bb4d0942fab099b69d7dc5

Author: Done in https://github.com/redhat-ai-dev/ai-rhdh-installer/pull/7/commits/5659a91ba4bc3ba06bb2b68130e890d0d1fa24e9

1. Fetch the dynamic plugins ConfigMap and save the `dynamic-plugins.yaml` content within to a temp file
    ```sh

>     kubectl get configmap backstage-dynamic-plugins-ai-rh-developer-hub -n ai-rhdh -o yaml | yq '.data["dynamic-plugins.yaml"]' > temp-dynamic-plugins.yaml

Reviewer: The namespace might defer from `ai-rhdh` right? So it might be better here to include something like `<your-namespace>`?

Reviewer 2: Yes I believe even in the ai-rhdh-installer case you can change the namespace you deploy the chart too so for clarity we should probably make this `<namespace>` instead of a hard `ai-rhdh`

Author: @Jdubrick @thepetk I decided to use the env var `$NAMESPACE` that refers to the rhdh namespace https://github.com/redhat-ai-dev/ai-rhdh-installer/pull/7/commits/b8753424110732cc94a6f2aec5575f72240e2392 as this step of setting it is highlighted at the beginning of each of these documents:
- [PIPELINES-CONFIG.md](https://github.com/michael-valdron/ai-rhdh-installer/blob/b8753424110732cc94a6f2aec5575f72240e2392/docs/PIPELINES-CONFIG.md?plain=1#L16)
- [GITOPS-CONFIG.md](https://github.com/michael-valdron/ai-rhdh-installer/blob/b8753424110732cc94a6f2aec5575f72240e2392/docs/GITOPS-CONFIG.md?plain=1#L19)

Reviewer: sgtm

    ```
2. Merge the contents of [`argocd-plugins.yaml`](../dynamic-plugins/argocd-plugins.yaml) into the temp file
    ```sh
    yq -i ".plugins += $(yq '.plugins' ./dynamic-plugins/argocd-plugins.yaml -M -o json) | .plugins |= unique_by(.package)" temp-dynamic-plugins.yaml
    ```
3. Patch the dynamic plugins ConfigMap with the updated content in the temp file
    ```sh
    kubectl patch configmap backstage-dynamic-plugins-ai-rh-developer-hub -n ai-rhdh \
    --type='merge' \
    -p="{\"data\":{\"dynamic-plugins.yaml\":\"$(echo "$(cat temp-dynamic-plugins.yaml)" | sed 's/"/\\"/g' | sed 's/$/\\n/g' | tr -d '\n')\"}}"
    ```

> 4. Dynamic plugins should be updated with the [ArgoCD plugins list](../dynamic-plugins/argocd-plugins.yaml) with a pod update triggered and you may remove the temp file at this point

Reviewer: ```suggestion
4. Dynamic plugins should be updated to include the [ArgoCD plugins list](../dynamic-plugins/argocd-plugins.yaml) with a pod update triggered and you may remove the temp file at this point
```

Reviewer: Just to avoid having "with" two times.

#### Step 3: Updating RHDH Deployment

> Now that all of the required ConfigMaps and Secrets are apart of the namespace, we must edit the RHDH Deployment to properly reference these items.

Reviewer: Might be my english here, just not quite sure if "are apart of" == "are part of" 
```suggestion
Now that all of the required ConfigMaps and Secrets are part of the namespace, we must edit the RHDH Deployment to properly reference these items.
```

Author: @thepetk Yeah I think this sounds better :+1:

### AI-RHDH-Installer: Manual Configuration

#### Step 1: ConfigMaps and Secrets
In your chosen namespace you should apply the `argocd-config` ConfigMap under [`argocd-config.yaml` ](../resources/argocd-config.yaml).

As part of the `ai-rhdh-installer` a secret was created in your desired namespace with the name `rhdh-argocd-secret`, keep note of this Secret as well as the two ConfigMaps applied above.

> #### Step 2.1: Updating Plugins with the web console

Reviewer 2: ```suggestion
#### Step 2.1: Updating Plugins Via Web Console
```

**Note: RHDH will encounter errors if the ArgoCD plugins are loaded before an instance is properly attached - This step may result in errors until all steps are completed**

To include the [ArgoCD plugins list](../dynamic-plugins/argocd-plugins.yaml) we need to edit the dynamic plugins ConfigMap that was created by the RHDH Operator:

![Dynamic Plugins Example](../assets/dynamic-plugins-example.png)

Edit the associated `yaml` file to include the contents of the [ArgoCD plugins list](../dynamic-plugins/argocd-plugins.yaml) under the `plugins` section:

![Dynamic Plugins Example 2](../assets/dynamic-plugins-example-2.png)

> #### Step 2.2: Updating Plugins with the CLI

Reviewer 2: ```suggestion
#### Step 2.2: Updating Plugins Via CLI
```

4. `ARGOCD_API_TOKEN`

#### Step 3: Updating Plugins

> You will follow the same steps as [step 2 for the ai-rhdh-installer](#step-2-updating-plugins).

Reviewer: ```suggestion
Follow the same steps as [step 2 for the ai-rhdh-installer](#step-2-updating-plugins).
```

#### Step 4: Updating RHDH Deployment

> Once you have applied the ConfigMaps and Secrets to your cluster and the necessary namespace you can now follow the same steps in [step 3 for the ai-rhdh-installer](#step-3-updating-rhdh-deployment). Every step will be identical if your RHDH instance was created using the `Red Hat Developer Hub Operator`, however, if you installed RHDH using the `Helm Chart` you may find that `.spec.template.spec.containers.envFrom` does not exist in the Deployment yaml. If this is the case you can simply add that field and the necessary secret information.

Reviewer: ```suggestion
Once you have applied the ConfigMaps and Secrets to your cluster and the chosen namespace you can now follow the same steps in [step 3 for the ai-rhdh-installer](#step-3-updating-rhdh-deployment). Every step will be identical if your RHDH instance was created using the `Red Hat Developer Hub Operator`, however, if you installed RHDH using the `Helm Chart` you may find that `.spec.template.spec.containers.envFrom` does not exist in the Deployment yaml. If this is the case you can simply add that field and the necessary secret information.
```

Reviewer: Might be better also here to be more detailed:

* `If this is the case you can simply add that field and the necessary secret information` 

I think maybe add this info as troubleshooting information might be a clearer approach? WDYT?

Author: > Might be better also here to be more detailed:
> 
>     * `If this is the case you can simply add that field and the necessary secret information`
> 
> 
> I think maybe add this info as troubleshooting information might be a clearer approach? WDYT?

@thepetk Yeah I agree, I think specifying `envFrom` again here would make it more clear, since there is a lot in this paragraph it made me think which field it was referring to as well. Also rewording to 'secret reference' rather than 'secret information'.

Author: Done in https://github.com/redhat-ai-dev/ai-rhdh-installer/pull/7/commits/b2898c501fd7b26a578ec9cdd15f7da7c58655bf
